### ISSUE TITLE
Wrong result with records and checkboxes

### ISSUE THREAD

**Issue by jugmac00** [zopefoundation/Zope#568]
When revising the "records arguments" section of the Zope developer book #567, I was reminded of an old bug, that bit me a couple of years ago.

Consider a form like this:

```
<form action="records_parse">

    <p>
    <input type="checkbox" name="index.enabled:records" value="0" />
    <input type="text" name="index.name:records" value="index 1" />
    <p>
    <input type="checkbox" name="index.enabled:records" value="1" checked="checked" />
    <input type="text" name="index.name:records" value="index 2" />
    <p>
    <input type="submit" name="submit" value="send" />

</form>
```
You'd expect a result as:
```
[{'name':'index 1'}, {'enabled':'1', 'name':'index 2'}]
```

But you get this
```
[{'enabled': '1', 'name': 'index 1'}, {'name': 'index 2'}]
```

So essentially the checkbox slips in the wrong list element.

This bug was reported already in 2010 by @petschki (hi!) on launchpad, and seemingly was forgotten to move to Github:
https://bugs.launchpad.net/zope2/+bug/619081

I can confirm that this bug is still present in the latest head of Zope 4.

As far as I understand the current test suite does not cover this bug as the test suite does not take the input types into account, and this bug only occurs with checkboxes.

As a workaround I used a hidden input field as the first input field as then the bug does not occour, in other words the bug only occurs, when the empty checkbox is the first input field of the form.

**Comment by d-maurer** [comment]
Jürgen Gmach wrote at 2019-4-26 21:35 -0700:
>When revising the "records arguments" section of the Zope developer book, I was reminded of an old bug, that bit me a couple of years ago.
>
>Consider a form like this:
>
>```
><form action="records_parse">
>
>    <p>
>    <input type="checkbox" name="index.enabled:records" value="0" />
>    <input type="text" name="index.name:records" value="index 1" />
>    <p>
>    <input type="checkbox" name="index.enabled:records" value="1" checked="checked" />
>    <input type="text" name="index.name:records" value="index 2" />
>    <p>
>    <input type="submit" name="submit" value="send" />
>
></form>
>```
>You'd expect a result as:
>```
>[{'name':'index 1'}, {'enabled':'1', 'name':'index 2'}]
>```
>
>But you get this
>```
>[{'enabled': '1', 'name': 'index 1'}, {'name': 'index 2'}]
>```
>
>So essentially the checkbox slips in the wrong list element.
>
>This bug was reported already in 2010 on launchpad, and seemingly was forgotten to move to Github:
>https://bugs.launchpad.net/zope2/+bug/619081
>
>I can confirm that this bug is still present in the latest head of Zope 4.

I fear this bug cannot be fixed: at the ZPublisher end it cannot be
recognized as a bug.

You can see that it is wrong because you see the input types
"checkbox" and "text" and you know that for a checkbox, the browser
sends the value only when the checkbox is checked.
The "ZPublisher" only sees a sequence of values named
"name", "enabled", "name". It has no information whether the
"enabled" should be paired with the first or the second "name".

For the form controls "<text><checkbox checked><text><checkbox>",
the "ZPublisher" would get the same sequence of name/values
and its result would be correct. But the "ZPublisher" does not
know about those form details.

I see only that the ZPublisher could raise an exception, issue a warning
or write a log record when a "records" variable get components with
inhomogenous fields - as in those cases, the ZPublisher might
have collected the individual name/value pairs in the wrong way.

**Comment by jugmac00** [comment]
Form with a hidden input field as a workaround
```
<form action="records_parse">

    <p>
    <input type="hidden" name="index.hidden:records" value="dummy" />
    <input type="checkbox" name="index.enabled:records" value="1" checked="checked" />
    <input type="text" name="index.name:records" value="index 1" />
    <p>
    <input type="hidden" name="index.hidden:records" value="dummy" />
    <input type="checkbox" name="index.enabled:records" value="2" />
    <input type="text" name="index.name:records" value="index 2" />
    <p>
    <input type="hidden" name="index.hidden:records" value="dummy" />
    <input type="checkbox" name="index.enabled:records" value="3" checked="checked" />
    <input type="text" name="index.name:records" value="index 3" />
    <p>
    <input type="submit" name="submit" value="send" />

</form>
```

Result as expected
```
[
{'enabled': '1', 'hidden': 'dummy', 'name': 'index 1'},
 {'hidden': 'dummy', 'name': 'index 2'},
 {'enabled': '3', 'hidden': 'dummy', 'name': 'index 3'}
]
```

@d-maurer Could you explain why this works? Does the publisher notice that when the second occurence of "hidden" appears it has to create a new record? Maybe even point to the source code - I have bit of a hard time finding the correct spot.

With the applied workaround the result is correct (or at least not mixed up and you can work with it), though it still contains inhomogenous fields. So I do not think raising a warning/exception is a proper solution.

If this bug really cannot be fixed, the documentation should to be updated with a "attention" paragraph.

**Comment by d-maurer** [comment]
Jürgen Gmach wrote at 2019-4-27 08:57 -0700:
> ...
>@d-maurer ... Maybe even point to the source code - I have bit of a hard time finding the correct spot.

Near line 736 of `ZPublisher.HTTPRequest` you find:


                    # Insert in dictionary
                    if key in mapping_object:
                        if flags & RECORDS:
                            # Get the list and the last record
                            # in the list. reclist is mutable.
                            reclist = mapping_object[key]
                            x = reclist[-1]
			    ...
                            if not hasattr(x, attr):
                                # If the attribute does not
                                # exist, setit
                                if flags & SEQUENCE:
                                    item = [item]
                                setattr(x, attr, item)
                            else:
                                if flags & SEQUENCE:
                                    # If the attribute is a
                                    # sequence, append the item
                                    # to the existing attribute
                                    y = getattr(x, attr)
                                    y.append(item)
                                    setattr(x, attr, y)
                                else:
                                    # Create a new record and add
                                    # it to the list
                                    n = record()
                                    setattr(n, attr, item)
                                    mapping_object[key].append(n)

This is essentially the code handling the ":records" type suffix
(I have omitted (--> "...") the convolution for the "tainted" handling).


After some head scratching, one notices that a new record is
started when the publisher sees again a field (name) for which the current
record already as a value and which is not a sequence.
In other words, a new record is started when the publisher cannot
put the name/value pair into the "current" record.

This approach is heuristical only.
It can fail for inhomogenous records (--> the bug report).


The publisher does not have sufficient information
to fill inhomogenous records reliably.

Compare the case from the bug report and the one from my previous comment.
In both cases, the publisher gets the same "name/value" pair sequence
(and therefore produces the same result).
In the first case, the result is wrong, in the second case correct.

**Comment by dataflake** [comment]
@jugmac00 as Dieter explained, this is a bug that can't really be fixed. While you're in the documentation I suggest you just add a warning box for that and close this issue.
